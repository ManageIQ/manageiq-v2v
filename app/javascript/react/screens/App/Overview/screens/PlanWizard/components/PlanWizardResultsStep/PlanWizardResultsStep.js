import React from 'react';
import PropTypes from 'prop-types';
import { noop, Icon } from 'patternfly-react';
import { planHasBeenEdited } from './helpers';
import WizardLoadingState from '../../../../../common/WizardLoadingState';
import WizardErrorState from '../../../../../common/WizardErrorState';

class PlanWizardResultsStep extends React.Component {
  componentDidMount() {
    const {
      postPlansUrl,
      postMigrationPlansAction,
      editPlansUrl,
      editMigrationPlansAction,
      plansBody,
      planSchedule,
      planType,
      editingPlan
    } = this.props;
    if (!editingPlan) {
      postMigrationPlansAction(postPlansUrl, plansBody, planSchedule, planType);
    } else if (planHasBeenEdited(editingPlan, plansBody, planSchedule)) {
      editMigrationPlansAction(editPlansUrl, editingPlan.id, plansBody, planSchedule, planType);
    }
  }
  renderResult = (migrationPlanMessage, migrationPlanFollowupMessage, migrationPlanIcon, showVmPowerWarning) => (
    <div className="wizard-pf-complete blank-slate-pf">
      <div className="modal-wizard-results-grey-icon">
        <span className={migrationPlanIcon} />
      </div>
      <h3 className="blank-slate-pf-main-action" id="migration-plan-results-message">
        {migrationPlanMessage}
      </h3>
      <p className="blank-slate-pf-secondary-action">{migrationPlanFollowupMessage}</p>
      {showVmPowerWarning && (
        <div className="plan-wizard-vm-power-warning">
          <Icon type="pf" name="warning-triangle-o" />
          <p>
            {__('VMs must be powered on in order to migrate.')}
            <br />
            {__('Ensure that all VMs in the Migration Plan are powered on before starting migration.')}
          </p>
        </div>
      )}
    </div>
  );

  render() {
    const {
      isPostingPlans,
      isRejectedPostingPlans,
      isPuttingPlans,
      isRejectedPuttingPlans,
      migrationPlansResult,
      migrationRequestsResult,
      errorPostingPlans,
      errorPuttingPlans,
      plansBody,
      planSchedule,
      planType,
      hidePlanWizardAction,
      targetProvider
    } = this.props;

    if (isPostingPlans) {
      return <WizardLoadingState title={__('Creating Migration Plan...')} />;
    } else if (isRejectedPostingPlans) {
      const errorData = errorPostingPlans && errorPostingPlans.data;
      const errorMessage = errorData && errorData.error && errorData.error.message;
      return (
        <WizardErrorState
          title={__('Error Creating Migration Plan')}
          message={errorMessage}
          onClose={hidePlanWizardAction}
        />
      );
    } else if (isPuttingPlans) {
      return <WizardLoadingState title={__('Saving Migration Plan...')} />;
    } else if (isRejectedPuttingPlans) {
      const errorData = errorPuttingPlans && errorPuttingPlans.data;
      const errorMessage = errorData && errorData.error && errorData.error.message;
      return (
        <WizardErrorState
          title={__('Error Saving Migration Plan')}
          message={errorMessage}
          onClose={hidePlanWizardAction}
        />
      );
    } else if (planType === 'migration_type_cold' && planSchedule === 'migration_plan_later' && migrationPlansResult) {
      const migrationPlanSaved = sprintf(__(" Migration Plan: '%s' has been saved"), plansBody.name);
      const migrationPlanFollowupMessage = __('Select Migrate on the Migration Plans page to begin migration');
      const showVmPowerWarning = targetProvider === 'openstack';
      return this.renderResult(
        migrationPlanSaved,
        migrationPlanFollowupMessage,
        'pficon pficon-ok',
        showVmPowerWarning
      );
    } else if (
      (planSchedule === 'migration_plan_now' || planType === 'migration_type_warm') &&
      migrationPlansResult &&
      migrationRequestsResult
    ) {
      const migrationPlanProgress = sprintf(__(" Migration Plan: '%s' is in progress"), plansBody.name);
      let migrationPlanFollowupMessage = null;
      if (planType === 'migration_type_cold') {
        migrationPlanFollowupMessage = __('This may take a long time. Progress of the plan will be shown in the Migration area'); // prettier-ignore
      } else if (planType === 'migration_type_warm') {
        migrationPlanFollowupMessage = __('Iterative copies of the VM data will begin immediately. Select an action on the Migration Plans page to schedule or begin the final cutover migration.'); // prettier-ignore
      }
      return this.renderResult(migrationPlanProgress, migrationPlanFollowupMessage, 'fa fa-clock-o');
    }
    return null;
  }
}
PlanWizardResultsStep.propTypes = {
  postPlansUrl: PropTypes.string,
  editPlansUrl: PropTypes.string,
  postMigrationPlansAction: PropTypes.func,
  editMigrationPlansAction: PropTypes.func,
  plansBody: PropTypes.object,
  planSchedule: PropTypes.string,
  planType: PropTypes.string,
  isPostingPlans: PropTypes.bool,
  isRejectedPostingPlans: PropTypes.bool,
  errorPostingPlans: PropTypes.object,
  isPuttingPlans: PropTypes.bool,
  isRejectedPuttingPlans: PropTypes.bool,
  errorPuttingPlans: PropTypes.object,
  migrationPlansResult: PropTypes.object,
  migrationRequestsResult: PropTypes.object,
  hidePlanWizardAction: PropTypes.func,
  editingPlan: PropTypes.object,
  targetProvider: PropTypes.string
};
PlanWizardResultsStep.defaultProps = {
  postPlansUrl: '/api/service_templates',
  editPlansUrl: '/api/service_templates',
  postMigrationPlansAction: noop,
  editMigrationPlansAction: noop,
  plansBody: {},
  planSchedule: '',
  planType: '',
  isPostingPlans: true,
  isRejectedPostingPlans: false,
  errorPostingPlans: null,
  migrationPlansResult: null,
  migrationRequestsResult: null,
  hidePlanWizardAction: noop
};
export default PlanWizardResultsStep;
