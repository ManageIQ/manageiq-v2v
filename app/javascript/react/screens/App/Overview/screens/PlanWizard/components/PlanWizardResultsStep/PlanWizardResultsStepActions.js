import API from '../../../../../../../../common/API';
import {
  POST_V2V_MIGRATION_PLANS,
  POST_V2V_MIGRATION_REQUESTS,
  PUT_V2V_MIGRATION_PLANS
} from './PlanWizardResultsStepConstants';

export { hidePlanWizardAction } from '../../PlanWizardActions';

const postMigrationRequestsAction = (response, dispatch) => {
  dispatch({
    type: POST_V2V_MIGRATION_REQUESTS,
    payload: new Promise((resolve, reject) => {
      API.post(`${response.data.results[0].href}`, {
        action: 'order'
      })
        .then(responseMigrationRequest => {
          resolve(responseMigrationRequest);
        })
        .catch(e => reject(e));
    })
  });
};

const _postMigrationPlansActionCreator = (url, migrationPlans, planSchedule, planType) => dispatch =>
  dispatch({
    type: POST_V2V_MIGRATION_PLANS,
    payload: new Promise((resolve, reject) => {
      API.post(url, migrationPlans)
        .then(response => {
          resolve(response);
          if (planSchedule === 'migration_plan_now' || planType === 'migration_type_warm') {
            postMigrationRequestsAction(response, dispatch);
          }
        })
        .catch(e => reject(e));
    })
  });

export const postMigrationPlansAction = (url, migrationPlans, planSchedule, planType) =>
  _postMigrationPlansActionCreator(url, migrationPlans, planSchedule, planType);

const _editMigrationPlansActionCreator = (url, planId, migrationPlans, planSchedule, planType) => dispatch => {
  const body = {
    action: 'edit',
    resource: { ...migrationPlans }
  };
  return dispatch({
    type: PUT_V2V_MIGRATION_PLANS,
    payload: new Promise((resolve, reject) => {
      API.post(`${url}/${planId}`, body)
        .then(response => {
          resolve(response);
          if (planSchedule === 'migration_plan_now' || planType === 'migration_type_warm') {
            postMigrationRequestsAction(response, dispatch);
          }
        })
        .catch(e => reject(e));
    })
  });
};

export const editMigrationPlansAction = (url, planId, migrationPlans, planSchedule, planType) =>
  _editMigrationPlansActionCreator(url, planId, migrationPlans, planSchedule, planType);
