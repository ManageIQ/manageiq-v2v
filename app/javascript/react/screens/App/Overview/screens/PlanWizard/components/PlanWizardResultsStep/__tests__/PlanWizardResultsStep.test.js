import React from 'react';
import { shallow } from 'enzyme';
import PlanWizardResultsStep from '../PlanWizardResultsStep';

describe('Plan wizard results step', () => {
  const getBaseProps = () => ({
    postPlansUrl: '/api/foo',
    editPlansUrl: '/api/bar',
    postMigrationPlansAction: jest.fn(),
    editMigrationPlansAction: jest.fn(),
    plansBody: { name: 'Mock Plan', mock: 'data' },
    planSchedule: 'migration_plan_later',
    isPostingPlans: false,
    isRejectedPostingPlans: false,
    errorPostingPlans: null,
    isPuttingPlans: false,
    isRejectedPuttingPlans: false,
    errorPuttingPlans: null,
    migrationPlansResult: { mock: 'data' },
    migrationRequestsResult: { mock: 'data' },
    hidePlanWizardAction: jest.fn(),
    editingPlan: null,
    targetProvider: 'openstack'
  });

  it('renders with a warning if migrating an OSP plan later', () => {
    const component = shallow(<PlanWizardResultsStep {...getBaseProps()} planType="migration_type_cold" />);
    expect(component).toMatchSnapshot();
  });

  it('renders with no warning if migrating an OSP plan immediately', () => {
    const component = shallow(
      <PlanWizardResultsStep {...getBaseProps()} planType="migration_type_cold" planSchedule="migration_plan_now" />
    );
    expect(component).toMatchSnapshot();
  });

  it('renders with info message if doing warm migration', () => {
    const component = shallow(<PlanWizardResultsStep {...getBaseProps()} planType="migration_type_warm" />);
    expect(component).toMatchSnapshot();
  });
});
