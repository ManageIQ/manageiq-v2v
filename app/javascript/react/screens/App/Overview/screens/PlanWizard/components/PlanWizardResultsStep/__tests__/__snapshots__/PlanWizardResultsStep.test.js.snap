// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plan wizard results step renders with a warning if migrating an OSP plan later 1`] = `
<div
  className="wizard-pf-complete blank-slate-pf"
>
  <div
    className="modal-wizard-results-grey-icon"
  >
    <span
      className="pficon pficon-ok"
    />
  </div>
  <h3
    className="blank-slate-pf-main-action"
    id="migration-plan-results-message"
  >
     Migration Plan: 'Mock Plan' has been saved
  </h3>
  <p
    className="blank-slate-pf-secondary-action"
  >
    Select Migrate on the Migration Plans page to begin migration
  </p>
  <div
    className="plan-wizard-vm-power-warning"
  >
    <Icon
      name="warning-triangle-o"
      type="pf"
    />
    <p>
      VMs must be powered on in order to migrate.
      <br />
      Ensure that all VMs in the Migration Plan are powered on before starting migration.
    </p>
  </div>
</div>
`;

exports[`Plan wizard results step renders with info message if doing warm migration 1`] = `
<div
  className="wizard-pf-complete blank-slate-pf"
>
  <div
    className="modal-wizard-results-grey-icon"
  >
    <span
      className="fa fa-clock-o"
    />
  </div>
  <h3
    className="blank-slate-pf-main-action"
    id="migration-plan-results-message"
  >
     Migration Plan: 'Mock Plan' is in progress
  </h3>
  <p
    className="blank-slate-pf-secondary-action"
  >
    Iterative copies of the VM data will begin immediately. Select an action on the Migration Plans page to schedule or begin the final cutover migration.
  </p>
</div>
`;

exports[`Plan wizard results step renders with no warning if migrating an OSP plan immediately 1`] = `
<div
  className="wizard-pf-complete blank-slate-pf"
>
  <div
    className="modal-wizard-results-grey-icon"
  >
    <span
      className="fa fa-clock-o"
    />
  </div>
  <h3
    className="blank-slate-pf-main-action"
    id="migration-plan-results-message"
  >
     Migration Plan: 'Mock Plan' is in progress
  </h3>
  <p
    className="blank-slate-pf-secondary-action"
  >
    This may take a long time. Progress of the plan will be shown in the Migration area
  </p>
</div>
`;
